### PR TITLE
refactor selector in drag service

### LINKED ISSUE #278: refactor selector in drag service

### PR DESCRIPTION
**Pull request by umarmohammed**
closes #278

### CODE REVIEW

File: projects/ngx-d3timeline/src/lib/content/activity-drag.service.ts [hunk 1]
```diff
@@ -1,37 +1,27 @@
...
 import * as fromActions from '../store/actions';
-import { ActivityRectangle } from './activity-rectangle';
 import { identifier } from '../types';
+import { selectDraggedToActivity } from './selectors/activity.selectors';
 
 @Injectable({ providedIn: 'root' })
 export class ActivityDragService {
   private dragEndSubject = new BehaviorSubject(null);
 
   drop$ = this.dragEndSubject.pipe(
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/content/activity-drag.service.ts]: something like `droppedActivity$`  or `droppedUpdatedActivity$` ? a bit hard to see exactly what this is to name it.

File: projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts [hunk 2]
```diff
@@ -46,8 +53,39 @@ const selectDraggedToSeries = createSelector(
...
+  );
+
 export const selectCurrentlyDraggedActivityWithDraggedToSeries = createSelector(
   selectCurrentlyDraggedActivity,
   selectDraggedToSeries,
   getCurrentlyDraggedActivityWithDraggedToSeries
 );
+
+export const selectDraggedToActivity = createSelector(
+  selectCurrentlyDraggedActivityWithDraggedToSeries,
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts]: Not sure of the need for this selector, its getting the dragged activity with the updated series, then the 2 selectors are getting updated times, then its put together. So one drag update field (series) is nested and the other two (times) aren't.

Why not have all the updates as selectors here then combine at the end?
```
selectCurrentlyDraggedActivity,
selectDraggedToSeries,
selectShiftedByDeltaTime(selectCurrentlyDraggedUpdatedStart),
selectShiftedByDeltaTime(selectCurrentlyDraggedUpdatedFinish),
updateActivityForDrag
```

File: projects/ngx-d3timeline/src/lib/drag-utils.ts [hunk 3]
```diff
@@ -45,41 +15,36 @@ export function getCurrentlyDraggedActivity(
...
-  timeOrientation: Orientation
-) {
-  const invert = getInverseBandScale(bandScale);
-  return dragEvent && timeOrientation === Orientation.Vertical
-    ? invert(dragEvent.x)
-    : invert(dragEvent.y);
-}
-
-function shiftedTimesForDraggingActivity(
+export function shiftUpdatedTimes(
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/drag-utils.ts]: as above this method may change, if it stays as is rename it something like `setUpdatedTimes` or `setUpdatedTimesForActivity`.

It is not shifting anything as current name implies.

File: projects/ngx-d3timeline/src/lib/drag-utils.ts [hunk 4]
```diff
@@ -45,41 +15,36 @@ export function getCurrentlyDraggedActivity(
...
+export function shiftTimeByRangeValue(
+  time: Date,
+  timeScale: TimeScale,
+  rangeValue: number
+): Date {
+  return timeScale.invert(timeScale(time) + rangeValue);
 }
 
-function getDeltaTime(
+export function getDeltaTime(
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/drag-utils.ts]: `getDeltaTimeForDrag` or `getTimeDeltaForDrag` might read better though we have been putting delta first elsewhere.
or even `getDragTimeDelta`

File: projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts [hunk 5]
```diff
@@ -46,8 +53,39 @@ const selectDraggedToSeries = createSelector(
...
+    shiftTimeByRangeValue
+  );
+
 export const selectCurrentlyDraggedActivityWithDraggedToSeries = createSelector(
   selectCurrentlyDraggedActivity,
   selectDraggedToSeries,
   getCurrentlyDraggedActivityWithDraggedToSeries
 );
+
+export const selectDraggedToActivity = createSelector(
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts]: maybe `selectActivityUpdatedForDrag'

File: projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts [hunk 6]
```diff
@@ -46,8 +53,39 @@ const selectDraggedToSeries = createSelector(
   valueToSeries
 );
 
+const selectDeltaTime = createSelector(
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts]: `selectTimeDeltaForDrag` + see other renaming suggestions of `getDeltaTime` further down.. the two should be consistent.

File: projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts [hunk 7]
```diff
@@ -46,8 +53,39 @@ const selectDraggedToSeries = createSelector(
...
+  selectTimeOrientation,
+  selectDragEvent,
+  getDeltaTime
+);
+
+const selectCurrentlyDraggedUpdatedStart = createSelector(
+  selectCurrentlyDraggedActivity,
+  activity => activity && activity.updatedStart
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts:64]: why do these need to return the activity? Tried to trace it through to where it's used and it looks like just the time is used?

File: projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts [hunk 8]
```diff
@@ -46,8 +53,39 @@ const selectDraggedToSeries = createSelector(
...
+);
+
+const selectCurrentlyDraggedUpdatedFinish = createSelector(
+  selectCurrentlyDraggedActivity,
+  activity => activity && activity.updatedFinish
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts:69]: as above

File: projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts [hunk 9]
```diff
@@ -46,8 +53,39 @@ const selectDraggedToSeries = createSelector(
...
+);
+
+const selectShiftedByDeltaTime = (selectTime: MemoizedSelector<Date>) =>
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts]: ```selectTimeShiftedForDragEvent``` or similar

ie: what is being shifted and why?

**Review by wflemingnz-retired — CHANGES_REQUESTED** [review]

File: projects/ngx-d3timeline/src/lib/content/activity-drag.service.ts (hunk 1 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/content/activity-drag.service.ts]: done

File: projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts (hunk 2 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts]: done

File: projects/ngx-d3timeline/src/lib/drag-utils.ts (hunk 3 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/drag-utils.ts]: done

File: projects/ngx-d3timeline/src/lib/drag-utils.ts (hunk 4 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/drag-utils.ts]: done

File: projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts (hunk 5 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts]: done

File: projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts (hunk 6 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts]: done

File: projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts (hunk 7 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts:64]: Its used in `selectActivityUpdatedForDrag`

File: projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts (hunk 8 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts:69]: Its used in `selectActivityUpdatedForDrag`

File: projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts (hunk 9 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts]: done

File: projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts (hunk 7 above)
Comment (wflemingnz-retired) [reply projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts:64]: In the code with the renamings it is used as an input to `selectShiftedByDeltaTime` where it is passed as a selector `selectTime`, the result of which is passed into the first parameter of `shiftTimeByRangeValue`. Note this parameter is a Date/time only, it does not use the Activity which comes back to my original question.

Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts:64]: Sorry I don't quite understand

> why do these need to return the activity? Tried to trace it through to where it's used and it looks like just the time is used?

They don't return the activity, they return a time value from an activity i.e. updatedStart or updatedEnd.

Comment (wflemingnz-retired) [reply projects/ngx-d3timeline/src/lib/content/selectors/activity.selectors.ts:64]: ah yes I see now, I was confusing `activity && activity.updatedStart` as a union type when its just a logical operator.

**Pull request merged**